### PR TITLE
User config

### PR DESCRIPTION
**Pull request by itpropro**
- Added parsing preparation for routes, navigationFallback. responseOverrides, globalHeaders and mimeTypes nodes
- Implemented response overrides redirection
- Added new useConfig type to wrap all config file nodes
- Fixed bug where the cli crashes on empty token
- Fixed redirection loop bug

### CODE REVIEW

**Comment by manekinekko** [comment]
Hey @itpropro thank you for opening this PR 🎉

Would you mind detailing what issue you're addressing? I can't find an open issue related to this PR?

Thank you.

File: src/proxy/routes-engine.ts [hunk 1]
```diff
@@ -3,34 +3,62 @@ import http from "http";
...
+  return config;
 };
 
 export const processUserRoute = async (req: http.IncomingMessage, res: http.ServerResponse, userDefinedRoutes: UserDefinedRoute[]) => {
   if (!req) {
     return Promise.resolve(-1);
   }
 
-  const userDefinedRoute = userDefinedRoutes.find((routeDef) => new RegExp(routeDef.route).test(req.url!));
+  const userDefinedRoute = userDefinedRoutes.find((routeDef) => new RegExp(`^${routeDef.route}`).test(req.url!));
```
Comment (itpropro) [inline src/proxy/routes-engine.ts:61]: Fix for recursive redirection, if path contains ".auth".

File: src/proxy/routes-engine.ts [hunk 2]
```diff
@@ -3,34 +3,62 @@ import http from "http";
 import { DEFAULT_CONFIG } from "../config";
 import { decodeCookie, findFile } from "../core/utils";
 
-export const handleUserCustomRoutes = async (folder: string): Promise<UserDefinedRoute[]> => {
+export const handleUserConfig = async (folder: string): Promise<UserConfig> => {
```
Comment (itpropro) [inline src/proxy/routes-engine.ts:6]: Refactored UserDefinedRoutes to UserConfig to handle all nodes inside the new SWA configuration file.

File: src/proxy/routes-engine.ts [hunk 3]
```diff
@@ -49,10 +77,14 @@ export const processUserRoute = async (req: http.IncomingMessage, res: http.Serv
     }
 
     // ACL
-    if (userDefinedRoute.allowedRoles && req.headers.cookie) {
-      const user = decodeCookie(req.headers.cookie);
-      console.log({ user });
-      if (!userDefinedRoute.allowedRoles.some((role) => user?.userRoles.some((ur: string) => ur === role))) {
+    if (userDefinedRoute.allowedRoles && !userDefinedRoute.allowedRoles.includes("anonymous")) {
```
Comment (itpropro) [inline src/proxy/routes-engine.ts:80]: Fixed problem, where the CLI crashed when req.headers.cookie is not set

File: src/proxy/server.ts [hunk 4]
```diff
@@ -37,12 +37,14 @@ if (!isHttpUrl(SWA_CLI_AUTH_URI)) {
...
 
+// make SWA assets referencable
 const SWA_404 = path.resolve(process.cwd(), "..", "public", "404.html");
 const SWA_401 = path.resolve(process.cwd(), "..", "public", "unauthorized.html");
+const htmlAssets: Record<string, string> = { SWA_404, SWA_401 };
 
 const serveStatic = (file: string, res: http.ServerResponse, status = 200) => {
   if (status !== 401) {
-    file = fs.existsSync(file) ? file : SWA_404;
+    file = fs.existsSync(file) ? file : htmlAssets["SWA_404"];
```
Comment (itpropro) [inline src/proxy/server.ts:47]: Added configuration object, to make the static responses dynamically addressable.

File: src/proxy/server.ts [hunk 5]
```diff
@@ -64,29 +66,29 @@ const injectClientPrincipalCookies = (cookie: string | undefined, proxyReqOrRes:
...
-        // ignore status -1
-        break;
+    // handle redirects
+    if (status) {
+      const override = userConfig.userDefinedResponseOverrides[`${status}`] || null;
+      if (override) {
+        res.writeHead(override.statusCode, { Location: override.redirect });
+      } else if ([401, 404].includes(status)) {
+        return serveStatic(htmlAssets[`SWA_${status}`], res, status);
+      }
```
Comment (itpropro) [inline src/proxy/server.ts:85]: Handle response overrides and reference static assets by dictionary key

File: src/swa.d.ts [hunk 6]
```diff
@@ -91,3 +91,30 @@ declare type UserDefinedRoute = {
...
+  "404": UserDefinedResponseOverride;
+};
+
+declare type UserConfig = {
+  userDefinedRoutes: UserDefinedRoute[];
+  userDefinedResponseOverrides?: Record<BadRequestOverride | UnauthorizedOverride | ForbiddenOverride | NotFoundOverride>;
+  navigationFallback?: Record<string, string>;
+  globalHeaders?: Record<string, string>;
+  mimeTypes?: Record<string, string>;
+};
```
Comment (itpropro) [inline src/swa.d.ts:120]: Added types for UserConfig
TODO: Currently only userDefinedRoutes and userDefinedResponseOverrides are handles in the application. The other three types need to be implemented and tested.

File: src/proxy/routes-engine.ts [hunk 7]
```diff
@@ -3,34 +3,62 @@ import http from "http";
 import { DEFAULT_CONFIG } from "../config";
 import { decodeCookie, findFile } from "../core/utils";
 
-export const handleUserCustomRoutes = async (folder: string): Promise<UserDefinedRoute[]> => {
+export const handleUserConfig = async (folder: string): Promise<UserConfig> => {
   if (!fs.existsSync(folder)) {
-    return [];
+    return { userDefinedRoutes: [] };
```
Comment (itpropro) [inline src/proxy/routes-engine.ts:8]: Return userConfig with an empty userDefinedRoutes array as that is the only required field and it's the same behavior as before.

File: readme.md [hunk 8]
```diff
@@ -167,6 +151,22 @@ When requesting the `.auth/me` endpoint, a `clientPrincipal` containing the fake
...
+- The Reverse Proxy: this is the heart of the SWA CLI, it's the piece that forwards all HTTP requests to the appropriate components:
+  - `/.auth/**` requests are forwarded to the Auth emulator server.
+  - `/api/**` requests are forwarded to the localhost API function (if available).
+  - `/**` all other requests are forwarded to the static assets server (serving the front-end app).
+- The Auth emulator server: emulates the whole authentication flow.
+- The Static content server: serves the local app static content.
+- The Serverless API server (served by the Azure Function App).
+
+Before SWA CLI bootstraps, it can also read (when using the `--build` options) the local SWA GitHub workflow file (created by [Azure Static Web Apps](https://docs.microsoft.com/azure/static-web-apps)) and builds both the static app and the API backend according based on the config. If the user isn't using an API backend, SWA CLI will skip the API backend build.
+
```
Comment (itpropro) [inline readme.md:169]: I didn't change anything here.

**Comment by itpropro** [comment]
> Hey @itpropro thank you for opening this PR 🎉
> 
> Would you mind detailing what issue you're addressing? I can't find an open issue related to this PR?
> 
> Thank you.

Hey,
I am actually developing on a few SWA applications and was trying to emulate functionality from the service. Where the emulator broke, I applied a fix. If you want, I can create separate issues for each of the problems I had.
The main issues were:
- Emulator crashed when there was no cookie existent ()
- Response overrides were not working
- Redirection was reclusively calling itself (RegEx was checking for .auth in the whole string not only at the beginning)

I added comments to all changes that I made in this PR, if you want me to create separate issues for each, I can do that as well.

**Comment by manekinekko** [comment]
Thank you @itpropro.

Yes, would you mind creating separate issues with their corresponding PRs? We are tracking individual issues to avoid having different people working on the same issue. You can find our contributing [guidelines here](https://github.com/Azure/static-web-apps-cli/blob/main/CONTRIBUTING.md#-submission-guidelines).

Thank you again for trying it the CLI. We really appreciate it ☺️

**Comment by itpropro** [comment]
Hey @manekinekko, 
no problem! Can you merge your PR (https://github.com/Azure/static-web-apps-cli/pull/86) then I can create issues for the things I have changed :)

**Comment by manekinekko** [comment]
I am sorry! That PR is not ready yet. I need to address Anthony's (and your) review first. Just make sure your rebase your changes on top of that branch.

It should be ready for merge before end of next week.

**Comment by manekinekko** [comment]
Hey @itpropro thank you again for creating this PR. I'd suggest we close it in favor of #86 

We'd also love to have your feedback on #86 in case you have the chance to try it out.

**Comment by manekinekko** [comment]
#86 has been merged. Closing this PR.

**Comment by itpropro** [comment]
#86 looks great, this one was just the result of me getting my demos to work. I just had no more time over the last days, but you implemented all I tried to do and much more 👍

**Pull request closed without merging**